File: JavaCountdown/PracticeRound/Scorer.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public class Scorer {

    static int lineSize(String line) {
        char[] chars = line.toCharArray();
        int size = 0;
        for (int i = 0; i < chars.length; i++) {
            if (!isSpace(chars[i]));
                size++;
        }
        return size;
    }

    private static boolean isSpace(char aChar) {
        return aChar == ' ';
    }

    static int linesSize(List<String> lines) {
        int size = 0;
        for (String line : lines) {
            size += lineSize(line);
        }
        return size;
    }


    static String hr(int width) {
        StringBuffer buf = new StringBuffer();

        buf.append("-----|");
        for (int i = 0; i < width; i++) {
            buf.append("-");
        }

        return buf.toString();
    }

    static void printProgramSize(List<String> lines) {


        int totalSize = 0;

        System.out.println(hr(60));

        for (String line : lines) {

            int size = lineSize(line);
            System.out.println("   " +  size + "  |" + line);
            totalSize += size;
        }

        System.out.println(hr(60));

        System.out.println("   " + totalSize + " == CountDown.java.size");

        System.out.println("");
    }

// - - - - - - - - - - - - - - - - - - - -

    static List<String> readLines(String filename) {
        List<String> lines = new ArrayList<>();

        try {
            BufferedReader reader = new BufferedReader(new FileReader(filename));
            String sCurrentLine;
            while ((sCurrentLine = reader.readLine()) != null) {
                lines.add(sCurrentLine);
            }
        } catch (IOException e) {
            throw new RuntimeException(e);
        }

        return lines;
    }

// - - - - - - - - - - - - - - - - - - - -

    static boolean uses(List<String> lines, String token) {
        // ... also matches .
        // double also matches do
        // etc etc

        for (String line : lines) {
            if (line.contains(token)) {
                return true;
            }
        }

        return false;
    }

    static int tokensSize(List<String> lines) {
        int size = 0;
        for (String token : Tokens.tokens) {
            if(uses(lines, token)) {
                size += token.length();
            }
        }

        return size;
    }

    static boolean missingTokens(List<String> lines) {
        List<String> unused = new ArrayList<>();
        for (String token : Tokens.tokens) {
            if (!uses(lines, token)) {
                return true;
            }
        }

        return false;
    }

// - - - - - - - - - - - - - - - - - - - -

    static void printTokenBonuses(List<String> lines) {
        int tokensSize = 0;
        System.out.println(hr(20));

        for (String token : Tokens.tokens) {
            if (uses(lines, token)) {
                System.out.println("   " + token.length() + "  |" + token);
                tokensSize += token.length();
            }
        }

        for (String token : Tokens.tokens) {
            if (!uses(lines, token)) {
                System.out.println("   " + 0 + "  |" + token);
            }
        }

        System.out.println(hr(20));

        System.out.println("   " + tokensSize + " == used_tokens.size");

        int completion_bonus = missingTokens(lines) ? 0 : 50;

        System.out.println("   " + completion_bonus + "  == completion.bonus");
    }

// - - - - - - - - - - - - - - - - - - - -

    public static void main(String[] args) {

        List<String> lines = readLines(args[1]);
        int program_size = linesSize(lines);
        int used_token_bonus = tokensSize(lines);
        int completion_bonus = missingTokens(lines) ? 0 : 50;

        System.out.println(">>> Score = -CountDown.java.size + 3*usedTokens.size + completion.bonus");
        System.out.println("   " + (-program_size) + " + 3*" + used_token_bonus + " + " + completion_bonus);
        System.out.println(">>>       = " + (-program_size + (3 * used_token_bonus) + completion_bonus));

        System.out.println();
        printTokenBonuses(lines);
        System.out.println();
        printProgramSize(lines);

        // green-traffic light pattern...put it out of sight
        for (int i = 0; i < 100; i++)
            System.out.println();
        System.out.println("All tests passed");
    }

}

File: RubyCountdown/Round2/scorer.rb
require "./tokens.rb"

class Scorer

    def line_size(line)
        size = 0
        line.each_char { |c|
            size += 1 unless is_space c
        }
        size
    end

    def is_space(char)
        char.strip.empty?
    end

    def lines_size(lines)
        size = 0
        lines.each { |line| size += line_size(line) }
        size
    end

    def hr(width)
        buf = ""
        buf << "------|"
        buf << ("-" * width)
    end

    def print_program_size(lines)

        totalsize = 0

        puts hr(60)

        lines.each do |line|
            size = line_size(line)
            printf(" %3d  |%s", size, line)
            totalsize += size
        end
        puts
        puts hr(60)

        puts "   " + totalsize.to_s + " == countdown.rb.size"
    end

    def read_lines(filename)
        lines = []
        f = File.open(filename, "r")
        f.each_line { |line|
            lines << line
        }
        f.close
        lines
    end

    def uses(lines, token)
        # ... also matches .
        # double also matches do
        # etc etc

        lines.each { |line| return true if (line.include? token) }

        false
    end

    def tokensSize(lines)
        size = 0

        TOKENS.each { |token|
             size += token.length if (uses(lines, token))
         }

        size
    end

    def missingTokens(lines)
        TOKENS.each { |token|
            return true if (!uses(lines, token))
        }

        false
    end

    def printTokenBonuses(lines)
        tokensSize = 0
        puts hr(20)

        TOKENS.each do |token|
            if (uses(lines, token)) then
                printf(" %3d  |%s\n", token.length(), token)
                tokensSize += token.length()
            end
        end

        TOKENS.each { |token|
            printf(" %3d  |%s\n", 0, token) if (!uses(lines, token))
        }

        puts hr(20)

        printf(" %3d == used_tokens.size\n", tokensSize)

        completion_bonus = missingTokens(lines) ? 0 : 50
        printf(" %3d == completion.bonus\n", completion_bonus)
    end

    def score_file(filename)
        score(read_lines(filename))

        # green-traffic light pattern...put it out of sight
        100.times { puts }
        puts "2 examples, 0 failures"
    end

    def score(lines)
        program_size = lines_size(lines)
        used_token_bonus = tokensSize(lines)
        completion_bonus = missingTokens(lines) ? 0 : 50

        puts ">>> Score = -CountDown.java.size + 5*usedTokens.size + completion.bonus"
        puts ">>>       = " + (-program_size).to_s + " + 5*" + used_token_bonus.to_s + " + " + completion_bonus.to_s
        puts ">>>       = " + (-program_size + (5 * used_token_bonus) + completion_bonus).to_s

        puts
        printTokenBonuses(lines)
        puts
        print_program_size(lines)
    end

end

Scorer.new.score_file("countdown.rb")
